docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Transfer Task (Activity Task)
+qp->>matching: enqueue Activity Task
+worker->>handler: RespondActivityTaskCompleted
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Transfer Task (Workflow Task)
+qp->>matching: enqueue Workflow Task
+worker->>handler: RespondWorkflowTaskCompleted(StartTimer)
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Timer Task
+qp->>qp: thirty days elapses<br><br>timer fired
+note right of handler: State transition:<br>append to Workflow Execution History<br>update Mutable State<br>create Transfer Task (Workflow Task)
+qp->>matching: enqueue Workflow Task
+```
+
+However, there are several requirements that make the above challenging to implement. In particular:
+
+- **Scaling**: the History service must be able to support arbitrarily many concurrent workflow executions. See [History shards](#history-shards).
+- **Consistency**: the Workflow and Activity tasks in Matching Service must be consistent with the workflow execution data in History Service. See [Consistency guarantees](#consistency-guarantees).
+- **Latency** of request handling and server-side workflow advance must be minimized.
+
+The rest of this document describes the implementation, which addresses these requirements.
+
+The History Service has many other responsibilities which are not covered here, for example: Shard Membership and Ownership, Replication and conflict resolution, Visibility, Archival, Workflow Versioning, Workflow Reset, Scheduled workflows.
+
+## Concepts
+
+### History shards
+
+A single Temporal Cluster manages many (e.g. millions of) Workflow Executions.
+These are logically partitioned into subsets that we call History Shards.
+The Temporal Cluster typically has multiple instances of the History service application process (each typically running on a separate machine/VM), and each History service instance owns some set of History Shards.
+"Owning" a History Shard means being responsible for the lifecycle of every workflow execution in the shard by [handling incoming requests](#rpc-handling) relating to that workflow execution, managing timers, and causing Workflow and Activity Tasks to be enqueued in Matching Service. To achieve this, each shard manages its own [task queues](#queue-processing).
+
+Therefore, objects corresponding to History Shards exist in the memory of a History service instance.
+Persistence details are abstracted away via a Go interface but in practice, when the persistence layer is backed by Cassandra, History Shards map 1-1 with partitions of the [`executions` table](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/schema/cassandra/temporal/schema.cql#L52). Persistence layer implementations also exist for MySQL, Postgres, and SQLite.
```

Comment:
Why referencing specific commit, not `main` branch? Yes, it can be broken, but link to inactive code is even worse.

Reply:
@flossypurse has a tool to manage links to the specific line of code in refactor proof way.

Reply:
> Why referencing specific commit, not `main` branch? Yes, it can be broken, but link to inactive code is even worse.

Links to `main` branch rapidly end up pointing at nonsense. The permalinks I'm using will always point to what they are intended to point to. So, it is guaranteed that the reader when following the link will get to the function, and it will be in a context where it made sense. So IMO if we include links with line numbers they should either be permalinks, or CI would need to guarantee validity of the link to `main`.

I'd still like to consider the possibility that we do include links with line numbers; I am finding them very useful. If it has moved on main then the reader will have a function name that they can search for in their IDE. But we would ideally like to detect when it has moved on main and when it has disappeared from main. This could in principle be done in CI.

Reply:
I'm going to merge this with the permalinks, but I agree it's not clear what the best way to do this is, so we can settle on our preferred approach and change if that's what we decide. Some options are

- permalinks with line numbers
- link to file only (in `main`), mentioning function/struct name
- no code links at all

In principle we could even use a format that permits a CI check to be written that confirms that the referenced name occurs on the referenced line. Or it might even be possible to automatically compute updated permalinks with sufficient accuracy, again in principle, not a high priority project.